### PR TITLE
Add files via upload

### PR DESCRIPTION
**Pull request by zch401**

### CODE REVIEW

**Comment by tjiang1898** [comment]
you could add .gitignore file to avoid check in the build artifacts such as .class files. See following for how the local gitignore works: https://help.github.com/en/github/using-git/ignoring-files#create-a-local-gitignore

File: proj1b/proj1b/Deque.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+    void addLast(T item);
+    int size();
+    void printDeque();
+    T removeFirst();
+    T removeLast();
+    T get(int index);
+    default boolean isEmpty() {
+        return size() == 0;
+    }
+}
```
Comment (tjiang1898) [inline proj1b/proj1b/Deque.java:12]: Looks good. It would look more professional if Javadoc comments are added so that one could generate API documentation via `javadoc` command. See https://www.baeldung.com/javadoc.

```
/**
 * Deque (usually pronounced like “deck”) is an irregular acronym of double-ended queue.
 * Double-ended queues are sequence containers with dynamic sizes that can be expanded 
 * or contracted on both ends (either its front or its back).
 **/
public interface Deque<T> {
    /**
     * Adds an item of type T to the front of the deque.
     * @param item: item to add
     **/
    void addFirst(T item);
```

File: proj1a/LinkedListDeque.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    public LinkedListDeque() {
+        sentinel = new LLDNode (null, null, null);
+        sentinel.prev = sentinel;
+        sentinel.next = sentinel;
+        size = 0;
+    }
+
+    /**add item to front of the deque*/
+    @Override
+    public void addFirst(T item) {
```
Comment (tjiang1898) [inline proj1a/LinkedListDeque.java:27]: You could make this implementation [thread safe](https://www.journaldev.com/1061/thread-safety-in-java) by adding `synchronized` keyword:

```
    public synchronized void addFirst(T item) {
```

File: proj1a/LinkedListDeque.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    public LinkedListDeque(LinkedListDeque other) {
+        sentinel = new LLDNode (null, null, null);
+        sentinel.prev = sentinel;
+        sentinel.next = sentinel;
+        size = 0;
+
+        for (int i = 0; i < other.size() ; i++ ) {
+            addLast((T) other.get(i));
+        }
+    }
```
Comment (tjiang1898) [inline proj1a/LinkedListDeque.java:125]: This code could be replaced with the default constructor:

```
    public LinkedListDeque(LinkedListDeque other) {
        this();
        for (int i = 0; i < other.size() ; i++ ) {
             addLast((T) other.get(i));
        }
    }
```

File: proj1a/LinkedListDeque.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    public T removeFirst() {
+        /**no first item*/
+        if (isEmpty()) {
+            return null;
+        }
+
+        T result = sentinel.next.item;
+        LLDNode target = sentinel.next.next;
+        LLDNode temp = new LLDNode(sentinel, target.item, target.next);
+        sentinel.next = temp;
```
Comment (tjiang1898) [inline proj1a/LinkedListDeque.java:77]: Don't quite understand the above code. Why adding another `LLDNode`? Shouldn't we remove the first `LLDNode` by making sure not reference to it? Would the number of `LLDNode` be inconsistent with `size`?

File: proj1a/LinkedListDeque.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        /**if index item does not exist*/
+        if (index > size - 1) {
+            return null;
+        }
+
+        int num = 0;
+        LLDNode p = sentinel.next;
+        while (num != index) {
+            p = p.next;
+            num = num + 1;
```
Comment (tjiang1898) [inline proj1a/LinkedListDeque.java:110]: You could use `num++` here

File: proj1a/ArrayDeque.java [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+    /**calculate nextFirst*/
+    private int nextF(int i) {
+        int next;
+        if (i == 0) {
+            next = items.length - 1;
+        } else {
+            next = i - 1;
+        }
+        return next;
```
Comment (tjiang1898) [inline proj1a/ArrayDeque.java:26]: Could be replaced with one liner as below
```
    private int nextF(int i) {
        return i == 0? items.length - 1 : i - 1
    }
```

File: proj1a/ArrayDeque.java [hunk 7]
```diff
@@ -0,0 +1,185 @@
+public class ArrayDeque<T> implements Deque<T> {
+    private T[] items;
+    private int size;
+    private int nextFirst;
+    private int nextLast;
+    private int RFACTOR;
```
Comment (tjiang1898) [inline proj1a/ArrayDeque.java:6]: Shouldn't this variable be initialized somewhere?

File: proj1a/ArrayDeque.java [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+    private int R;
```
Comment (tjiang1898) [inline proj1a/ArrayDeque.java:7]: Why do we need this instance variable?

File: proj1a/ArrayDeque.java [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+    /**resize array, create a larger one*/
+    private void resizeL(int capacity) {
+        T[] temp = (T[]) new Object[capacity];
+        int old_size = size;
+        size = 0;
+        nextFirst = 4;
+        nextLast = 5;
+        for (int i = 0; i < old_size ; i++ ) {
+            addFirst((T) get(i));
```
Comment (tjiang1898) [inline proj1a/ArrayDeque.java:48]: How does this line of code add items to the `temp` array? Where is the code unit that unit tests this function?

File: proj1a/ArrayDeque.java [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+    /**resize array, create a smaller one*/
+    private void resizeS(int capacity) {
+        T[] temp = (T[]) new Object[capacity];
+        int old_size = size;
+        size = 0;
+        nextFirst = 4;
+        nextLast = 5;
+        for (int i = 0; i < old_size ; i++ ) {
+            addFirst((T) get(i));
```
Comment (tjiang1898) [inline proj1a/ArrayDeque.java:61]: Could you add unit test to see if this code actually works?

File: proj1a/ArrayDeque.java [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+        size =  size - 1;
+        nextLast = pos;
+        return result;
+    }
+
+    /**get item at given index*/
+    @Override
+    public T get(int index) {
+        /**if the index does not exist*/
+        if (index >= size ) {
```
Comment (tjiang1898) [inline proj1a/ArrayDeque.java:160]: What about `index < 0`?

File: proj1a/ArrayDeque.java [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+
+        if (pos >= items.length ) {
+            pos = pos - items.length;
+        }
+        return items[pos];
+    }
+
+    /**create a deep copy*/
+    public ArrayDeque(ArrayDeque other) {
+        items = (T[]) new Object[8];
```
Comment (tjiang1898) [inline proj1a/ArrayDeque.java:176]: it would be more efficient if you initialize the array to `other.items.length`

**Comment by tjiang1898** [comment]
Could you add unit tests for `ArrayDeque` as I see some bugs in it?

File: proj1a/LinkedListDequeTest.java [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+		boolean passed = checkEmpty(true, lld1.isEmpty());
+
+		lld1.addFirst(10);
+		// should not be empty
+		passed = checkEmpty(false, lld1.isEmpty()) && passed;
+
+		lld1.removeFirst();
+		// should be empty
+		passed = checkEmpty(true, lld1.isEmpty()) && passed;
+
```
Comment (tjiang1898) [inline proj1a/LinkedListDequeTest.java:82]: Should add some negative tests such as removing from empty list.

File: proj1a/ArrayDeque.java (hunk 8 above)
Comment (zch401) [reply proj1a/ArrayDeque.java:7]: This is used to represent usage ratio of the array. However, this should not be an instance for the ArrayDeque class. I have put the definition of this variable into the corresponding method where it was used.

File: proj1a/ArrayDeque.java (hunk 11 above)
Comment (zch401) [reply proj1a/ArrayDeque.java:160]: case index < 0, return null added in the new code

File: proj1a/ArrayDeque.java (hunk 8 above)
Comment (tjiang1898) [reply proj1a/ArrayDeque.java:7]: Yes, it should be a local variable
